File: trunk/csharp/core/GenericWithPresentationFactory.cs
using System;
using System.Collections.Generic;
using System.Reflection;
using System.IO;
using System.Xml;
using urakawa.exception;
using urakawa.xuk;

namespace urakawa
{
    /// <summary>
    /// Generic base class for creation of instances of types implementing <see cref="WithPresentation"/>
    /// </summary>
    /// <typeparam name="T">The base <see cref="Type"/> of instances created by the factory</typeparam>
    public abstract class GenericWithPresentationFactory<T> : GenericXukAbleFactory<T> where T : WithPresentation
    {
        protected override void XukInChild(XmlReader source, progress.ProgressHandler handler)
        {
            if (source.LocalName == XukStrings.RegisteredTypes && source.NamespaceURI == XukNamespaceUri)
            {
                XukInRegisteredTypes(source, handler);
            }
            else if (!Presentation.Project.IsPrettyFormat())
            {
                XukInRegisteredType(source);
            }

            //we're replacing the super class method completely
            //base.XukInChild(source, handler);
        }
        protected override void XukOutChildren(XmlWriter destination, Uri baseUri, progress.ProgressHandler handler)
        {
            if (Presentation.Project.IsPrettyFormat())
            {
                destination.WriteStartElement(XukStrings.RegisteredTypes, XukNamespaceUri);
            }

            XukOutRegisteredTypes(destination, baseUri, handler);

            if (Presentation.Project.IsPrettyFormat())
            {
                destination.WriteEndElement();
            }
            //we're replacing the super class method completely
            //base.XukOutChildren(destination, baseUri, handler);
        }

        public override bool IsPrettyFormat()
        {
            return Presentation.IsPrettyFormat();
        }

        public override void SetPrettyFormat(bool pretty)
        {
            Presentation.SetPrettyFormat(pretty);
        }

        private Presentation mPresentation;

        /// <summary>
        /// Gets the <see cref="Presentation"/> associated with <c>this</c>
        /// </summary>
        /// <returns>The owning <see cref="Presentation"/></returns>
        public Presentation Presentation
        {
            get
            {
                return mPresentation;
            }
        }

        public GenericWithPresentationFactory(Presentation pres)
        {
            mPresentation = pres;
        }

        /// <summary>
        /// Inistalizes an created instance by assigning it an owning <see cref="urakawa.Presentation"/>
        /// </summary>
        /// <param name="instance">The instance to initialize</param>
        /// <remarks>
        /// In derived factories, this method can be overridden in order to do additional initialization.
        /// In this case the developer must remember to call <c>base.InitializeInstance(instance)</c>
        /// </remarks>
        protected override void InitializeInstance(T instance)
        {
            base.InitializeInstance(instance);
            instance.Presentation = Presentation;
        }
    }
}

File: trunk/java/core/src/org/daisy/urakawa/GenericWithPresentationFactory.java
package org.daisy.urakawa;

import org.daisy.urakawa.exception.IsAlreadyInitializedException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;

/**
 * @param <T>
 * @depend - Aggregation 1 org.daisy.urakawa.Presentation
 */
public abstract class GenericWithPresentationFactory<T extends AbstractXukAbleWithPresentation>
        extends GenericXukAbleFactory<T>
{
    private Presentation mPresentation;

    /**
     * @return the Presentation owner
     */
    public Presentation getPresentation()
    {
        return mPresentation;
    }

    /**
     * @param pres
     * @throws MethodParameterIsNullException
     */
    public GenericWithPresentationFactory(Presentation pres)
            throws MethodParameterIsNullException
    {
        if (pres == null)
        {
            throw new MethodParameterIsNullException();
        }
    }

    /**
     * Initializes a created instance by assigning it an Presentation owner. In
     * derived classes of this factory, this method can be overridden in order
     * to perform additional initialization, in which case
     * super.InitializeInstance(instance) must be called.
     */
    @Override
    protected void initializeInstance(T instance)
    {
        super.initializeInstance(instance);
        try
        {
            instance.setPresentation(getPresentation());
        }
        catch (MethodParameterIsNullException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        catch (IsAlreadyInitializedException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
    }
}
